File: src/main/java/org/svomz/apps/finances/domain/model/TransactionRepository.java
package org.svomz.apps.finances.domain.model;

import java.util.List;
import java.util.Optional;
import java.util.Set;

/**
 * Created by eric on 15/04/17.
 */
public interface TransactionRepository {

  void create(Transaction expense);

  TransactionId nextIdentity();

  List<Transaction> findByAccountId(AccountId accountId, int page, int size);

  List<Transaction> findAllByAccountId(AccountId accountId);

  Optional<Transaction> findById(AccountId accountId, TransactionId transactionId);

  void update(Transaction transaction);

  void delete(Transaction transaction);

  List<Transaction> findAllByTag(AccountId accountId, Tag tag);

  Set<Tag> findAllTags(AccountId accountId);
}

File: core/src/main/scala/org/svomz/apps/finances/core/domain/model/TransactionRepository.scala
package org.svomz.apps.finances.core.domain.model

import java.util.Date

import scala.concurrent.Future

trait TransactionRepository {

  def nextIdentity: Future[TransactionId]

  def create(expense: Transaction): Future[Transaction]

  def update(transaction: Transaction): Future[Transaction]

  def delete(transaction: Transaction): Future[Transaction]

  def fetch(accountId: AccountId, transactionId: TransactionId): Future[Option[Transaction]]

  def fetchAll(accountId: AccountId): Future[List[Transaction]]

  def fetchAllDebits(id: AccountId, from: Some[Date], to: Some[Date]): Future[List[Transaction]]

  def fetchAllCredits(id: AccountId, from: Some[Date], to: Some[Date]): Future[List[Transaction]]

  def fetchAllWithCategory(accountId: AccountId, category: Category, from: Option[Date], to: Option[Date]): Future[List[Transaction]]

  def fetchAllCategories(accountId: AccountId):  Future[List[Category]]

  def fetchAllDebitWithCategory(id: AccountId, c: Category, from: Option[Date], to: Option[Date]): Future[List[Transaction]]


}
